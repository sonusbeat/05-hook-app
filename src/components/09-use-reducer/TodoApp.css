.description {
  cursor: pointer;
  margin: 0;
}

.completed {
  text-decoration: line-through;
}