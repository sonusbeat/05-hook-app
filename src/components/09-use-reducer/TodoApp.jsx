import { useReducer } from "react";
import TodoReducer from './todoReducer';
import "./TodoApp.css"

const initialState = [{
  id: new Date().getTime(),
  description: "Learn React",
  done: false
}];

const TodoApp = () => {

  const [ todos, dispatch ] = useReducer(TodoReducer, initialState);

  const handleSubmit = (event) => {
    event.preventDefault();

    const newTodo = {
      id: new Date().getTime(),
      description: "New Task",
      done: false
    };

    const todoAction = {
      type: "add",
      payload: newTodo
    }

    dispatch(todoAction);
  };

  console.log(todos);

  return (
    <section>
      <h2 className="text-center mb-4">Todo App <small>({ todos.length })</small></h2>

      <div className="mb-5">
        <form onSubmit={ handleSubmit }>
          <div className="form-group">
            <input
              name="description"
              type="text"
              className="form-control"
              autoComplete="off"
            />
          </div>
          <button
            type="submit"
            className="btn btn-primary btn-block"
          >Agregar</button>
        </form>
      </div>

      <table className="table">
        <thead>
          <tr>
            <th>Task</th>
            <th className="text-center">Done</th>
            <th className="text-center">Actions</th>
          </tr>
        </thead>
        <tbody>
          {
            todos.map((todo, i) => (
              <tr key={todo.id}>
                <td>
                  <p className="description">{ `${i + 1}. ${todo.description}` }</p>
                </td>
                <td className="text-center">{ todo.done ? "Yes" : "No" }</td>
                <td className="text-center">
                  <button className="btn btn-danger btn-sm">Delete</button>
                </td>
              </tr>
            ))
          }
        </tbody>
      </table>
    </section>
  );
}

export default TodoApp;