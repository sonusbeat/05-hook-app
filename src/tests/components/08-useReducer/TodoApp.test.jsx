import TodoApp from '../../../components/09-use-reducer/TodoApp';
import { shallow, mount } from 'enzyme';
import demoTodos from '../../fixtures/demotTodos';
import { act } from '@testing-library/react';

describe("Pruebas en <TodoApp />", () => {

  const wrapper = shallow( <TodoApp /> );

  Storage.prototype.setItem = jest.fn(() => {});

  test('Debe de coincidir con el snapshot', () => {
    expect( wrapper ).toMatchSnapshot();
  });

  test('Debe de agregar dos TODO', () => {

    const wrapper = mount( <TodoApp />);

    act(() => {
      wrapper.find("TodoAdd").prop("HandleAddTodo")(demoTodos[0]);
      wrapper.find("TodoAdd").prop("HandleAddTodo")(demoTodos[1]);
    });

    expect(wrapper.find("h2").text().trim()).toBe("Todo App (2)");
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);

  });

  test('Debe de eliminar un TODO', () => {

    wrapper.find("TodoAdd").prop("HandleAddTodo")( demoTodos[0] );
    wrapper.find("TodoList").prop("HandleDelete")( demoTodos[0].id );
    expect(wrapper.find("h2").text().trim()).toBe("Todo App (0)");

  });

});