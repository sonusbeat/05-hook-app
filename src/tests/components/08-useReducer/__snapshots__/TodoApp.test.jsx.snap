// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> Debe de coincidir con el snapshot 1`] = `
<section>
  <h2
    className="text-center mb-4"
  >
    Todo App 
    <small>
      (
      0
      )
    </small>
  </h2>
  <TodoAdd
    HandleAddTodo={[Function]}
  />
  <div
    className="alert alert-warning text-center"
  >
    <strong>
      AÃºn no hay tareas creadas
    </strong>
  </div>
</section>
`;
